compat/src/main/scala-2.11_2.12/scala/collection/compat/CompatImpl.scala

```diff
@@ -12,10 +12,47 @@
 
 package scala.collection.compat
 
+import scala.reflect.ClassTag
 import scala.collection.generic.CanBuildFrom
 import scala.collection.mutable.Builder
 import scala.collection.{immutable => i, mutable => m}
 
+/* builder optimized for a single ++= call, which returns identity on result if possible
+ * and defers to the underlying builder if not.
+ */
+private final class IdentityPreservingBuilder[A, CC[X] <: TraversableOnce[X]](that: Builder[A, CC[A]])(implicit ct: ClassTag[CC[A]])
+    extends Builder[A, CC[A]] {
+  var collection: CC[A] = null.asInstanceOf[CC[A]]
+  var ruined = false
+
+  final override def ++=(elems: TraversableOnce[A]): this.type =
+      elems match {
+        case ct(ca) if (collection == null && !ruined) =>  {
```

Comment:
I think you can just do `case ca: CC[A] if (...) =>` and the compiler figures it out

Reply:
Unfortunately not. I suggest removing the `ClassTag` parameter but taking instead the partial function that performs the runtime test, as a parameter.

Reply:
@julienrf are you sure? With this format tests pass (though start failing when I go to Nth's proposal). What's the failure mode?

Reply:
I guess it just doesn't work with HKT. It works fine with normal kinded types:

```scala
Welcome to Scala 2.12.8 (OpenJDK 64-Bit Server VM, Java 1.8.0_212).
Type in expressions for evaluation. Or try :help.

scala> def test[A: reflect.ClassTag](elem: Any): Boolean = elem match {
     |   case a: A => true
     |   case _ => false
     | }
test: [A](elem: Any)(implicit evidence$1: scala.reflect.ClassTag[A])Boolean

scala> test[Int]("foo")
res0: Boolean = false

scala> test[Int](1)
res1: Boolean = true

scala> test[String](1)
res2: Boolean = false

scala> test[String]("foo")
res3: Boolean = true
```

Reply:
Wow interesting, TIL that matching on a universally quantified type `A` does work if there is an implicit `ClassTag` in scope.

However, I suggest doing the following instead, I’m always unsure how reliable are `ClassTag`s on Scala.js.

~~~ scala
class IdentityPreservingBuilder[A, CC[X] <: TraversableOnce[X]](
  that: m.Builder[A, CC[A]],
  extractCC: PartialFunction[TraversableOnce[A], CC[A]]
) extends m.Builder[A, CC[A]] {

  private val extracted = new PartialFunctionAsExtractor(extractCC) // Unfortunately, #7111 is not available on 2.12 and 2.11

...

  override def ++=(elems: TraversableOnce[A]): this.type =
    elems match {
      case extracted(ca) if (collection == null && !ruined) =>
        collection = ca
        this
      case _ =>
        ruin()
        that ++= elems
        this
    }

...

}

private class PartialFunctionAsExtractor[CC[_], A](pf: PartialFunction[TraversableOnce[A], CC[A]]) {
  def unapply(elems: TraversableOnce[A]): Option[CC[A]] = pf.lift(elems)
}
~~~

And then:

~~~ diff
- case c.Seq | i.Seq => new IdentityPreservingBuilder[A, i.Seq](i.Seq.newBuilder[A])
+ case c.Seq | i.Seq =>
+   new IdentityPreservingBuilder[A, i.Seq](i.Seq.newBuilder[A], { case xs: i.Seq[A] => xs })
~~~

Reply:
isn't that pretty much exactly what `ClassTag[i.Seq[_]]#unapply.unlift` is?

@sjrd could you shine your light on classtag reliability?

Reply:
> isn't that pretty much exactly what `ClassTag[i.Seq[_]]#unapply.unlift` is?

Yes, that’s equivalent.

Reply:
The non-deprecated parts of `ClassTag`s work in Scala.js exactly as on the JVM.

The deprecated method `<:<`, which really is a method of the `Manifest` API, doesn't link on Scala.js.